Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -171,6 +231,52 @@ theorem bitwise'_comm {f : Bool → Bool → Bool} (hf : ∀ b b', f b b' = f b'
     _ = swap (bitwise' f) := bitwise'_swap hf'
 #align nat.bitwise_comm Nat.bitwise'_comm
 
+lemma bitwise_eq_bitwise' (f: Bool → Bool → Bool ) (h: f false false = false) :
```

Comment:
Did you have a counterexample that demonstrated why we need `h`? This was discussed on Zulip, right?

Reply:
Yes. From what I understood, we don't need ```h```. However, without it, it's hard to prove...

Reply:
I think we should:
* Prove that they are equal
* Remove one of them, unless there's a good reason to have both definitions.

Reply:
I've been trying to prove this lemma without `h`, but it is indeed quite hard.

Both `bitwise` and `bitwise'` already exist in master, and a decent number of existing `bitwise'` theorems carry a similar assumption, so I'd argue that this current PR does not worsen the current status quo and I don't think we should block it on this issue.

(For future reference: the zulip thread mentioned is: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Merging.20bitwise.20and.20bitwise')

Reply:
Yes, I'm trying to prove this too and it is indeed quite challenging.

Reply:
Can I make the following suggestions for how to prove it. The issue seems to boil down to the fact that in a sense, `binaryRec` defines the case `n=0` twice, because `bit false 0 = 0`.

I think there should be a version of `binaryRec_eq` with the different hypothesis `n = 0 -> b = true`

```
theorem binaryRec_eq' {C : Nat → Sort u} {z : C 0} {f : ∀ b n, C n → C (bit b n)}
    (b n) (hbn : n = 0 → b = true) : binaryRec z f (bit b n) = f b n (binaryRec z f n)
```

Then the theorem `bitwise'_bit` can be stated as 
```
theorem bitwise'_bit' {f : Bool → Bool → Bool} (a m b n) 
    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
    bitwise' f (bit a m) (bit b n) = bit (f a b) (bitwise' f m n) :=
```
And theorems could even be proven with a stronger version of `binaryRec`.
```
def binaryRec' {C : Nat → Sort u} (z : C 0) (f : ∀ b n, (n = 0 → b = true) → C n → C (bit b n)) : ∀ n, C n :=
```

Reply:
Thanks! It turns out `Nat.binaryRec_eq'` and `binaryRec'` already exist in `Data.Nat.Bits`. With the former, I finished and pushed a proof of `bitwise_eq_bitwise'` that doesn't need the extra assumption